packages/nuxt/src/core/nitro.ts

```diff
@@ -517,24 +520,54 @@ export async function initNitro (nuxt: Nuxt & { _nitro?: Nitro }) {
     })
   }
 
+  async function symlinkDist () {
+    if (nitro.options.static) {
+      const distDir = resolve(nuxt.options.rootDir, 'dist')
+      if (!existsSync(distDir)) {
+        await fsp.symlink(nitro.options.output.publicDir, distDir, 'junction').catch(() => {})
+      }
+    }
+  }
+
   // nuxt build/dev
   nuxt.hook('build:done', async () => {
     await nuxt.callHook('nitro:build:before', nitro)
     if (nuxt.options.dev) {
-      await build(nitro)
-    } else {
-      await prepare(nitro)
-      await prerender(nitro)
-
-      logger.restoreAll()
-      await build(nitro)
-      logger.wrapAll()
+      return build(nitro)
+    }
 
-      if (nitro.options.static) {
-        const distDir = resolve(nuxt.options.rootDir, 'dist')
-        if (!existsSync(distDir)) {
-          await fsp.symlink(nitro.options.output.publicDir, distDir, 'junction').catch(() => {})
+    const foldersToCache = {
+      dir: nitro.options.output.dir,
+      server: nitro.options.output.serverDir,
+      public: nitro.options.output.publicDir,
+    }
+    if (nuxt.options.experimental.buildCache) {
+      const { hash: cacheId } = await getNitroHash(nuxt)
+      const cacheDir = join(nuxt.options.rootDir, 'node_modules/.cache/nuxt/builds/nitro', cacheId)
+      if (existsSync(cacheDir)) {
+        for (const [key, folder] of Object.entries(foldersToCache)) {
+          await cp(join(cacheDir, key), folder, { recursive: true })
         }
+
+        await symlinkDist()
+        consola.success('Restored Nuxt Nitro server from cache.')
+        return
+      }
+    }
+    await prepare(nitro)
+    await prerender(nitro)
+
+    logger.restoreAll()
+    await build(nitro)
+    logger.wrapAll()
+
+    await symlinkDist()
+
+    if (nuxt.options.experimental.buildCache) {
+      const { hash: cacheId } = await getNitroHash(nuxt)
+      const cacheDir = join(nuxt.options.rootDir, 'node_modules/.cache/nuxt/builds/nitro', cacheId)
+      for (const [key, folder] of Object.entries(foldersToCache)) {
+        await cp(folder, join(cacheDir, key), { recursive: true })
```

Comment:
@pi0 do you think we need to/should use `nanotar` rather than copying the folders?

Reply:
random read/write on filesystem is costly this is for perf (tar is literally a buffer of files so unless random access to cache files is needed i think this is faster)